src/controller/tests/TestReadChunking.cpp

```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
     emberAfClearDynamicEndpoint(0);
 }
 
+/*
+ * This test contains two parts, one is to enable a new endpoint on the fly, another is to disable it and re-enable it.
+ */
+void TestCommandInteraction::TestDynamicEndpoint(nlTestSuite * apSuite, void * apContext)
+{
+    TestContext & ctx                    = *static_cast<TestContext *>(apContext);
+    auto sessionHandle                   = ctx.GetSessionBobToAlice();
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+
+    // Initialize the ember side server logic
+    InitDataModelHandler(&ctx.GetExchangeManager());
+
+    // Register our fake dynamic endpoint.
+    DataVersion dataVersionStorage[ArraySize(testEndpoint4Clusters)];
+
+    app::AttributePathParams attributePath;
+    app::ReadPrepareParams readParams(sessionHandle);
+
+    readParams.mpAttributePathParamsList    = &attributePath;
+    readParams.mAttributePathParamsListSize = 1;
+    readParams.mMaxIntervalCeilingSeconds   = 1;
+
+    TestReadCallback readCallback;
+
+    {
+
+        app::ReadClient readClient(engine, &ctx.GetExchangeManager(), readCallback.mBufferedCallback,
+                                   app::ReadClient::InteractionType::Subscribe);
+
+        NL_TEST_ASSERT(apSuite, readClient.SendRequest(readParams) == CHIP_NO_ERROR);
+
+        for (int j = 0; j < 3 && !readCallback.mOnSubscriptionEstablished; j++)
+        {
+            ctx.DrainAndServiceIO();
+            chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+            ctx.DrainAndServiceIO();
+        }
+
+        // We should not receive any reports in initial reports, so check mOnSubscriptionEstablished instead.
+        NL_TEST_ASSERT(apSuite, readCallback.mOnSubscriptionEstablished);
+        readCallback.mAttributeCount = 0;
+
+        // Enable the new endpoint
+        emberAfSetDynamicEndpoint(0, kTestEndpointId4, &testEndpoint4, 0, 0, Span<DataVersion>(dataVersionStorage));
+
+        for (int j = 0; j < 2 && !readCallback.mOnReportEnd; j++)
+        {
+            ctx.DrainAndServiceIO();
+            chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+            ctx.DrainAndServiceIO();
+        }
```

Comment:
Again, there should be no manual `Run()` call and just a single `ctx.DrainAndServiceIO();` here.
```suggestion
        ctx.DrainAndServiceIO();
```